src/internal/lokiutil/testloki/logger.go

```diff
@@ -0,0 +1,138 @@
+package testloki
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/lokiutil/client"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachconfig"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachd"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/randutil"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func WithTestLoki(l *TestLoki) pachd.TestPachdOption {
+	return pachd.TestPachdOption{
+		MutateEnv: func(env *pachd.Env) {
+			env.GetLokiClient = func() (*client.Client, error) {
+				return l.Client, nil
+			}
+		},
+		MutateConfig: func(config *pachconfig.PachdFullConfiguration) {
+			config.LokiLogging = true
+		},
+		MutateContext: func(ctx context.Context) context.Context {
+			templateHash := randutil.UniqueString("")[0:10]
+			procHash := randutil.UniqueString("")[0:5]
+			return pctx.Child(ctx, "", l.WithLoki(ctx, map[string]string{
+				"host":              "localhost",
+				"app":               "pachd",
+				"container":         "pachd",
+				"node_name":         "localhost",
+				"pod":               fmt.Sprintf("pachd-%v-%v", templateHash, procHash),
+				"pod_template_hash": templateHash,
+				"stream":            "stderr",
+				"suite":             "pachyderm",
+			}))
+		},
+	}
+}
+
+// newZapCore returns a zapcore.Core that sends logs to this Loki instance.
+func (l *TestLoki) newZapCore(ctx context.Context) *lokiCore {
+	return &lokiCore{
+		ctx: ctx,
+		l:   l,
+		enc: zapcore.NewJSONEncoder(zapcore.EncoderConfig{
+			TimeKey:        "time",
+			EncodeTime:     zapcore.RFC3339NanoTimeEncoder,
+			LevelKey:       "severity",
+			EncodeLevel:    zapcore.LowercaseLevelEncoder,
+			MessageKey:     "message",
+			NameKey:        "logger",
+			CallerKey:      "caller",
+			FunctionKey:    zapcore.OmitKey,
+			StacktraceKey:  "stacktrace",
+			LineEnding:     zapcore.DefaultLineEnding,
+			EncodeDuration: zapcore.SecondsDurationEncoder,
+			EncodeCaller:   zapcore.ShortCallerEncoder,
+		}),
+		labels: map[string]string{"suite": "pachyderm"},
+	}
+}
+
+// WithLoki returns a pctx context option that will send logs to this Loki instance.
+func (l *TestLoki) WithLoki(sendCtx context.Context, lokiLabels map[string]string) pctx.Option {
```

Comment:
Yes, the context containing the logger is not the one we use to do HTTP requests to loki.  Sending to loki is normally a background task handled by another process running on the machine, but now we're building it into our code, so we need some context to mediate the HTTP requests.   I think with WithLoki API is fine; pctx.Child takes ...pctx.Option, this is one more you can add to it to send logs to Loki.  The context taken by this function controls which context is used for this background task.  The context that the option mutates controls how logging happens in that block.

Reply:
But does it need to be exported? Nothing other than `testloki.WithTestLoki` calls `testloki.WithLoki`, so can the latter be private — or are you planning to expand its use in the future?

Reply:
Yeah, I think it can be exported.  If you are writing a test without testpachd, which is totally reasonable, now you can send logs to Loki.  That could be useful for something.

Reply:
> That could be useful for something.

That’s fair. And in that case, the caller would want to provide his own labels, not the ones created by the un-called testpachd-related `WithTestLoki`.